File: src/examples/USZipCodeMultipleLookupsExample.cs
﻿namespace Examples
{
	using System;
	using System.IO;
    using System.Net;
    using SmartyStreets;
	using SmartyStreets.USZipCodeApi;

	internal static class USZipCodeMultipleLookupsExample
	{
		public static void Run()
		{
			// specifies the TLS protocoll to use - this is TLS 1.2
			const SecurityProtocolType tlsProtocol1_2 = (SecurityProtocolType)3072;

			// You don't have to store your keys in environment variables, but we recommend it.
			var authId = Environment.GetEnvironmentVariable("SMARTY_AUTH_ID");
			var authToken = Environment.GetEnvironmentVariable("SMARTY_AUTH_TOKEN");
			ServicePointManager.SecurityProtocol = tlsProtocol1_2;

			using var client = new ClientBuilder(authId, authToken)
				.BuildUsZipCodeApiClient();

			var lookup1 = new Lookup
			{
				InputId = "dfc33cb6-829e-4fea-aa1b-b6d6580f0817", // Optional ID from your system
				ZipCode = "12345"
			};

			//uncomment the lines below to add custom parameters
			// lookup1.AddCustomParameter("city", "Schenectady");
			// lookup1.AddCustomParameter("state", "NY");

			var lookup2 = new Lookup
			{
				City = "Phoenix",
				State = "Arizona"
			};

			var lookup3 = new Lookup("cupertino", "CA", "95014") // You can also set these with arguments
			{
				InputId = "01189998819991197253"
			};

			var batch = new Batch();

			try
			{
				batch.Add(lookup1);
				batch.Add(lookup2);
				batch.Add(lookup3);

				client.Send(batch);
			}
			catch (BatchFullException)
			{
				Console.WriteLine("Error. The batch is already full.");
				return;
			}
			catch (SmartyException ex)
			{
				Console.WriteLine(ex.Message);
				Console.WriteLine(ex.StackTrace);
				return;
			}
			catch (IOException ex)
			{
				Console.WriteLine(ex.StackTrace);
				return;
			}

			for (var i = 0; i < batch.Count; i++)
			{
				var result = batch[i].Result;
				Console.WriteLine("Lookup " + i + ":\n");

				if (result.Status != null)
				{
					Console.WriteLine("Status: " + result.Status);
					Console.WriteLine("Reason: " + result.Reason);
					continue;
				}

				Console.WriteLine("Input ID: " + result.InputId);

				var cityStates = result.CityStates;
				var zipCodes = result.ZipCodes;

				if (cityStates == null || zipCodes == null)
				{
					Console.WriteLine("Lookup " + i + " is invalid.\n");
					continue;
				}

				Console.WriteLine(cityStates.Length + " City and State match" + (cityStates.Length == 1 ? ":" : "es:"));

				foreach (var cityState in cityStates)
				{
					Console.WriteLine("City: " + cityState.City);
					Console.WriteLine("State: " + cityState.State);
					Console.WriteLine("Mailable City: " + cityState.MailableCity);
					Console.WriteLine();
				}

				Console.WriteLine(zipCodes.Length + " ZIP Code match" + (cityStates.Length == 1 ? ":" : "es:"));

				foreach (var zipCode in zipCodes)
				{
					Console.WriteLine("ZIP Code: " + zipCode.ZipCode);
					Console.WriteLine("County: " + zipCode.CountyName);
					Console.WriteLine("Latitude: " + zipCode.Latitude);
					Console.WriteLine("Longitude: " + zipCode.Longitude);
					Console.WriteLine();
				}

				Console.WriteLine("***********************************");
			}
		}
	}
}

File: src/vbexamples/USZipCodeMultipleLookupsExample.vb
Imports System.Reflection.Emit
Imports System.Windows
Imports System.IO
Imports SmartyStreets
Imports SmartyStreets.USZipCodeApi
Imports System.Formats

Module USZipCodeMultipleLookupsExample

	Sub USZipCodeMultipleLookupsExample()

		Dim authID = Environment.GetEnvironmentVariable("SMARTY_AUTH_ID")
		Dim authToken = Environment.GetEnvironmentVariable("SMARTY_AUTH_TOKEN")

		Dim client = New ClientBuilder(authID, authToken).BuildUsZipCodeApiClient()
		Dim batch = New Batch()

		Dim lookup0 As New Lookup()
		With lookup0
			.InputId = "dfc33cb6-829e-4fea-aa1b-b6d6580f0817"
			.ZipCode = "12345"
		End With

		Dim lookup1 As New Lookup
		With lookup1
			.City = "Phoenix"
			.State = "Arizona"
		End With

		Dim lookup2 As New Lookup("cupertino", "CA", "95014")
		With lookup2
			.InputId = "01189998819991197253"
		End With

		Console.WriteLine("*******************************************************")
		Console.WriteLine()

		Try
			batch.Add(lookup0)
			batch.Add(lookup1)
			batch.Add(lookup2)
			Client.Send(batch)
		Catch ex As BatchFullException
			Console.WriteLine("Error. The batch is already full.")
		Catch ex As SmartyException
			Console.WriteLine(ex.Message)
			Console.WriteLine(ex.StackTrace)
		Catch ex As IOException
			Console.WriteLine(ex.StackTrace)
		Catch ex As Exception
			Console.WriteLine(ex.Message)
			Console.WriteLine(ex.StackTrace)
		End Try

		Dim numLookups = batch.Count

		For i As Integer = 0 To numLookups - 1
			If i > 0 Then
				Console.WriteLine()
			End If

			Dim result = batch(i).Result
			Dim cities = result.CityStates
			Dim zipCodes = result.ZipCodes

			If result.Status IsNot Nothing Then
				Console.WriteLine("Lookup " + CStr(i) + " has an invalid status.")
				Console.WriteLine("Status: " + result.Status)
				Console.WriteLine("Reason: " + result.Reason)
				Continue For
			End If

			If zipCodes Is Nothing Or cities Is Nothing Then
				Console.WriteLine("Lookup " + CStr(i) + " has no candidates. The lookup is not valid.")
				Continue For
			End If

			Console.WriteLine("Lookup " + CStr(i) + " is valid." + Environment.NewLine())

			Console.WriteLine("Input ID: " + result.InputId + Environment.NewLine())

			Console.WriteLine(CStr(cities.Length) + " city and state match" + If(cities.Length = 1, ":", "es:"))

			For Each cityState In cities
				Console.WriteLine()
				Console.WriteLine("City: " + cityState.City)
				Console.WriteLine("State: " + cityState.State)
				Console.WriteLine("Mailable City: " + CStr(cityState.MailableCity))
			Next

			Console.WriteLine()
			Console.WriteLine(CStr(zipCodes.Length) + " ZIP code match" + If((zipCodes.Length = 1), ":", "es:"))

			For Each zipCode In zipCodes
				Console.WriteLine()
				Console.WriteLine("ZIP Code: " + zipCode.ZipCode)
				Console.WriteLine("County: " + zipCode.CountyName)
				Console.WriteLine("Latitude: " + CStr(zipCode.Latitude))
				Console.WriteLine("Longitude: " + CStr(zipCode.Longitude))
			Next

		Next

		Console.WriteLine()

	End Sub

End Module
